### PR TITLE
vchiq: Missing commits from downstream

### PR DESCRIPTION
**Pull request by popcornmix**
Just tested vchiq in 4.9 and it works fine with downstream kernel.
I notices some commits present in 4.4 tree were missing.

### CODE REVIEW

**Comment by anholt** [comment]
Note: I'm not allowed to merge pull requests; all commits must go through the email workflow.  I'm barely keeping up with reviewing/acking vchiq patches as they're sent in, and I don't have time to manage being the one forwarding this set upstream.

**Comment by lategoodbye** [comment]
AFAIK these patches hasn't been merged yet. I could take care of them, but they aren't ready for submission yet.

**Review by lategoodbye — CHANGES_REQUESTED** [review]
This needs a commit message which explain why this is necessary.

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c [hunk 1]
```diff
@@ -891,16 +891,14 @@ queue_message(VCHIQ_STATE_T *state, VCHIQ_SERVICE_T *service,
...
-					if (SRVTRACE_ENABLED(service,
-							VCHIQ_LOG_INFO))
-						vchiq_log_dump_mem("Sent", 0,
-							header->data + pos,
-							min(64u,
-							elements[0].size));
-				}
 			}
 
+		if (SRVTRACE_ENABLED(service,
```
Comment (lategoodbye) [inline drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c:900]: This needs a commit message which explain why this complete patch is necessary.

**Comment by popcornmix** [comment]
Are you referring to the first logging tweak commit, or something else?

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c [hunk 2]
```diff
@@ -210,18 +210,26 @@ add_completion(VCHIQ_INSTANCE_T instance, VCHIQ_REASON_T reason,
 	VCHIQ_COMPLETION_DATA_T *completion;
 	DEBUG_INITIALISE(g_state.local)
 
+	mutex_lock(&instance->completion_mutex);
```
Comment (lategoodbye) [inline drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c]: If i get it right, the following patch reverts this patch, so this should be dropped completely.

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c [hunk 3]
```diff
@@ -64,10 +64,10 @@
 #define VCHIQ_MINOR 0
 
 /* Some per-instance constants */
-#define MAX_COMPLETIONS 16
+#define MAX_COMPLETIONS 128
```
Comment (lategoodbye) [inline drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c:67]: The commit messages doesn't explain why it's increased. So this should be a separate patch.

**Comment by lategoodbye** [comment]
Yes. Sorry, i'm using the review feature for the first time and i didn't find a delete button.

**Comment by popcornmix** [comment]
@pelwell can you provide a suitable commit comment for the logging tweak commit?

**Review by lategoodbye — CHANGES_REQUESTED** [review]
Here are my notes before i'm able to send the patches upstream.

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c (hunk 1 above)
Comment (pelwell) [reply drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c:900]: * Move the trace message outside the gather loop so fragmentation
doesn't mask significant bytes.
* Reduce the size of the data logged - 16 bytes is sufficient.

Signed-off-by: Phil Elwell <phil@raspberrypi.org>

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c (hunk 3 above)
Comment (lategoodbye) [reply drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c:67]: @pelwell Thanks for the other comment, but the changes to MAX_COMPLETIONS and MSG_QUEUE_SIZE needs a separate commit message, too.

Comment (pelwell) [reply drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c:67]: vchiq_arm: Avoid premature stalls

The constants MAX_COMPLETIONS and MSG_QUEUE_SIZE control the number of
messages that can be outstanding to each client before the system as a
whole stalls. If the numbers are too small then unnecessary thread
switching will occur while waiting for a (potentially low priority)
client thread to consume some data; badly written clients can even
lead to deadlock.

For services that carry many short messages, 16 messages can represent
a very small amount of data. Since the resources are small - 16 bytes
for a completion, 4 bytes for a message pointer -  increase the limits
so they are unlikely to be hit except in exceptional circumstances.

Signed-off-by: Phil Elwell <phil@raspberrypi.org>

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c [hunk 4]
```diff
@@ -610,22 +610,24 @@ process_free_queue(VCHIQ_STATE_T *state)
...
+	mb();
+
 	while (slot_queue_available != local->slot_queue_recycle) {
 		unsigned int pos;
 		int slot_index = local->slot_queue[slot_queue_available++ &
 			VCHIQ_SLOT_QUEUE_MASK];
 		char *data = (char *)SLOT_DATA_FROM_INDEX(state, slot_index);
 		int data_found = 0;
 
+		rmb();
```
Comment (lategoodbye) [inline drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c:629]: @pelwell I've rebased and rearranged this series. But checkpatch complains about a missing comment why this read memory barrier is necessary.

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c [hunk 5]
```diff
@@ -741,6 +743,8 @@ process_free_queue(VCHIQ_STATE_T *state)
 				up(&state->data_quota_event);
 		}
 
+		mb();
```
Comment (lategoodbye) [inline drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c:746]: @pelwell checkpatch complains here, too

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c (hunk 1 above)
Comment (lategoodbye) [reply drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c:900]: Thanks. It seems that the first part has already been applied.

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c (hunk 5 above)
Comment (pelwell) [reply drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c:746]:
```
		/* Don't allow the slot to be reused until we are no
		** longer interested in it. */
```

File: drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c (hunk 4 above)
Comment (pelwell) [reply drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.c:629]: /* Beware of the address dependency - data is calculated
		** using an index written by the other side. */

**Comment by Electron752** [comment]
This is just a heads up that I submitted a set of patches for the ioctl compatibility cleanup without realizing that these changes modify some of the same areas.  It looks like it's mostly the await for completion ioctl that has a conflict.

I don't know what set may get applied first...

**Comment by lategoodbye** [comment]
Yes, i didn't expect such a patch set one day after mine. I will add you to CC next time.

Btw real ARM64 test results for this [patch series](http://driverdev.linuxdriverproject.org/pipermail/driverdev-devel/2017-January/098977.html) are appreciated.

**Comment by Electron752** [comment]
Sure thing on the testing. I'm going to be away from the computer for awhile, so I won't be able to get to test it until later in the day.

It doesn't look very likely to break ARM64 though but I'll test it anyway.

**Comment by lategoodbye** [comment]
Thanks, but no need to hurry. In best case the patch set gets applied next week.

**Comment by Electron752** [comment]
OK, I ran some tests on arm64 with the patches and it appears to work.

I ran multiple instances of arm64 vchiq_test concurrently running "-f 200" and "-p 5000".  As far as I can tell nothing is broken.

It is a bit difficult to tests the upstream tree well of course since none of the downstream drivers that use vchiq are in the tree.

**Comment by lategoodbye** [comment]
Thanks again. I think it's not a big problem having no vchiq user driver upstream as long as the downstream drivers could use it.

**Comment by lategoodbye** [comment]
The patch series has been merged to [staging-next](https://git.kernel.org/cgit/linux/kernel/git/gregkh/staging.git/log/?h=staging-next), so i think this pull request could be closed.

**Comment by popcornmix** [comment]
Thanks!
